src/hud.cpp

```diff
@@ -196,24 +196,25 @@ void Hud::drawItem(const ItemStack &item, const core::rect<s32>& rect,
 	}
 
 //NOTE: selectitem = 0 -> no selected; selectitem 1-based
-void Hud::drawItems(v2s32 upperleftpos, s32 itemcount, s32 offset,
-		InventoryList *mainlist, u16 selectitem, u16 direction)
+void Hud::drawItems(v2s32 upperleftpos, s32 itemcount, s32 inv_offset,
+		InventoryList *mainlist, u16 selectitem, u16 direction, const v2s32 &offset)
 {
 #ifdef HAVE_TOUCHSCREENGUI
-	if ( (g_touchscreengui) && (offset == 0))
+	if (g_touchscreengui && inv_offset == 0)
 		g_touchscreengui->resetHud();
 #endif
 
 	s32 height  = m_hotbar_imagesize + m_padding * 2;
-	s32 width   = (itemcount - offset) * (m_hotbar_imagesize + m_padding * 2);
+	s32 width   = (itemcount - inv_offset) * (m_hotbar_imagesize + m_padding * 2);
 
 	if (direction == HUD_DIR_TOP_BOTTOM || direction == HUD_DIR_BOTTOM_TOP) {
 		width  = m_hotbar_imagesize + m_padding * 2;
-		height = (itemcount - offset) * (m_hotbar_imagesize + m_padding * 2);
+		height = (itemcount - inv_offset) * (m_hotbar_imagesize + m_padding * 2);
 	}
 
 	// Position of upper left corner of bar
-	v2s32 pos = upperleftpos;
+	v2s32 pos = upperleftpos + offset;
+	pos *= g_settings->getFloat("hud_scaling") * porting::getDisplayDensity();
```

Comment:
Should probably cache hud_scaling in the constructor, since it's already assumed to be unmodified throughout the Hud object lifetime.